### ISSUE TITLE
The typings on `marshall` are too narrow

### ISSUE THREAD

**Issue by vivmaha** [aws/aws-sdk-js-v3#1647]
**Describe the bug**
`marshall` from `@aws-sdk/util-dynamodb` only accepts data in the form of 

````
{
    [key: string]: NativeAttributeValue;
}
````

This narrow typing prevents you from using `marshall` with:
1) typescript `interface`s. If you use a typescript `type` it works.
2) arrays. The underlying `convertToAttr` properly supports this, so why not expose it up?

The v2 sdk's `DocumentClient` handled both of these cases, so I'm guessing `marshall` should do the same if parity is a factor.

**SDK version number**

1.0.0-rc.2

**Is the issue in the browser/Node.js/ReactNative?**
Node.js

**Details of the browser/Node.js/ReactNative version**
12

**To Reproduce (observed behavior)**

````
interface MyType = {
  foo: string;
  nested: {
    baz: string;
  };
};

const mySimpleObject: MyType = {
  foo: "foo",
  nested: {
    baz: "baz",
  },
};

const x = marshall(mySimpleObject);
^^^^ Argument of type 'MyType' is not assignable to parameter of type '{ [key: string]: NativeAttributeValue; }'.
  Index signature is missing in type 'MyType'.ts(2345)

````

**Expected behavior**
I should be able to pass in `interface` types into `marshall`. I should be able to also pass in `[]`s.

**Screenshots**

**Additional context**

**Comment by trivikr** [comment]
@vivmaha The DocumentClient in v2 had several bugs. For example, it stored a Date object as a map.
The types in `marshall` are narrower to avoid bugs like these.

We're exploring an option to provide premarshall configuration in https://github.com/aws/aws-sdk-js-v3/issues/1533#issuecomment-702178924

However, the interface support might not require a premarshall configuration and can be stored as a map. Marking this as feature request.

**Comment by stang-tgs** [comment]
The marshaller seems to be very selective on what it deems to be an object 😄 .  It can't handle any custom classes and will fail on any named classes.  Perhaps the marshaller should be doing `typeof` checks instead of constructor names?  Or a combination of?  However, this will have the same side effects as the v2 sdk.  The bright side is that JS Sets exist now, so potentially that can be marshalled.  JS `Maps` may be weird, because they don't look very different than `Objects` to Dynamo. 🤔 

In any case, the marshaller is a little useless for custom objects at the moment; I'd say it's more of bug than feature because of the super narrowness for objects `data?.constructor?.name === "Object"`

![image](https://user-images.githubusercontent.com/32852813/102437916-2fa4a980-3fd0-11eb-98fd-c2f7a77129aa.png)

**Comment by trivikr** [comment]
For reference: This [TypeScript Playground](https://www.typescriptlang.org/play?target=7#code/JYOwLgpgTgZghgYwgAgHIFcC2AjaA1OAG3RQG8AoZZKCOAEwHsRCBPZANyJIC5kBnMFFABzANzkAvuXJgWABxSo4YYOwgBBMIODZ0kAsRQBeSsgA+aZaogBlBEThRN23fq4RTF0sgDaAawgWXgEhEGEAXV4lFTVnIVcIAxJkKSoLaOs4nT1E9x9wz2QbCDAAHhAsXChzZGxgYVAwGowcfHcakJFmq1iteJyAIVBHNgt0EDoIGFAIOgA+cRl5RR7be0JHLISk40KKwkJC8cnpkFnC7AYGQloQPcroQpaqnYv6xqfVrcHhqBZCzphRayBSWGIaPrZSBDEAjZAmNLIdRQKBwFgDdAwGCPREDQgMbCFAAiyjgeGAEAA7oUAGLAG6FACS4AAHMjUf9EQBVRpslFowo81kAYQ2mAUdHZAsRzLAAEYAGxSzk1IXypX8lUWWUAZgATMrBY19YbETT8coTZraRawAqACymmpDYSyh1OiwutXuzWLRrQeBIZAAWRYABVlsgKFQYFdgtogaYzgJZrxo1RanAAF7x0JiUwScSF6QIJgCZCYFg2YDim4AeWwACsIAgwLxQxHQUYo6ZYwxeAAiPsDgA0SYgKboadMVGw2cHc6zo4LY+L5FLIHLmEcfAAFkRCMy5Ho074AkF+AmIlEvpDtu0JPCK1Wa3J602W2BREA) shows the Type Error when passing Interface as an input to marshall function discussed in this issue.

**Comment by trivikr** [comment]
I asked [the question in StackOverflow](https://stackoverflow.com/questions/65832224) for a solution to create a generic TypeScript interface which matches any type/interface/object with restrictions on types for values inside it.

While writing the question, I noticed that existing setup accepts types instead of interfaces as an input for marshall.

<details>
<summary>Example code</summary>

```ts
import { DynamoDB, waitForTableExists } from "@aws-sdk/client-dynamodb";
import { marshall } from "@aws-sdk/util-dynamodb";

type TableItemType = {
  id: string;
};

(async () => {
  const region = "us-west-2";
  const client = new DynamoDB({ region, logger: console });

  const TableName = `test-${Math.ceil(Math.random() * 10 ** 10)}`;

  const params = {
    TableName,
    AttributeDefinitions: [{ AttributeName: "id", AttributeType: "S" }],
    KeySchema: [{ AttributeName: "id", KeyType: "HASH" }],
    BillingMode: "PAY_PER_REQUEST",
  };

  await client.createTable(params);
  await waitForTableExists({ client, maxWaitTime: 60 }, { TableName });

  const itemContents: TableItemType = { id: "id" };
  await client.putItem({
    Item: marshall(itemContents),
    TableName,
  });

  await client.getItem({
    Key: marshall(itemContents),
    TableName,
  });

  await client.deleteTable({ TableName });
})();

```

</details>

**Comment by trivikr** [comment]
After going through [StackOverflow Answer](https://stackoverflow.com/a/65832777/477683), and issue https://github.com/microsoft/TypeScript/issues/15300 on TypeScript repo, the recommendation is to either:
* Use spread operator while passing interface object to marshall.
* Use Type instead of Interface. 

Verified that spread operator works:

<details>
<summary>Example code</summary>

```ts
import { DynamoDB, waitForTableExists } from "@aws-sdk/client-dynamodb";
import { marshall } from "@aws-sdk/util-dynamodb";

interface TableItemType {
  id: string;
}

(async () => {
  const region = "us-west-2";
  const client = new DynamoDB({ region, logger: console });

  const TableName = `test-${Math.ceil(Math.random() * 10 ** 10)}`;

  const params = {
    TableName,
    AttributeDefinitions: [{ AttributeName: "id", AttributeType: "S" }],
    KeySchema: [{ AttributeName: "id", KeyType: "HASH" }],
    BillingMode: "PAY_PER_REQUEST",
  };

  await client.createTable(params);
  await waitForTableExists({ client, maxWaitTime: 60 }, { TableName });

  const itemContents: TableItemType = { id: "id" };
  await client.putItem({
    Item: marshall({ ...itemContents }),
    TableName,
  });

  await client.getItem({
    Key: marshall({ ...itemContents }),
    TableName,
  });

  await client.deleteTable({ TableName });
})();

```

</details>

Verified that TypeScript will still throw error is Interface contains values which are not allowed.
[TypeScript Playground](https://www.typescriptlang.org/play?target=7#code/JYOwLgpgTgZghgYwgAgLIE8CS5ryQdWDAAsBlMKUAcwDU4AbAVwgGdkBvAKGWRgHs+mAApQ+ABwBcyFhWoBubsgBGcKMNGTpskFQUBfBZwR8QM5MEIlyUKRmyRYiCJbLbaDZmwC8HRf0Ei4lIARP7BADSKKmqBmsjB0cGcBpycoA54KHY4jgRErlAAgiAAJgByjAC2vjwyMRpSdfKKIFXqQcitlUrQ+obGpmDmLtbF5VW2WDmZIxRjFdU+XLUU7ZrBdREtbbFSAIzJhum4TmjoAOIQINDACPYnSBzIANoA1hDojW4Aul+UOsg9KkAPTA5BlPjIaCiKBGExmCz5azYMSMMCTS7XSh3aanJbIAB0RMRVgogMMoOQABVSMgAKJQGFSWLQMDoZAAci6aw55jYoGMlTEcDAwCU9BQAHd8uZShAAB7SYBUEAixhQCAEuGDYZIualBYotEYq43HEZPFPIkEkkFeZVclAA)

**Comment by stang-tgs** [comment]
FYI - in our ported code, we already are using the spread operator before marshalling, to sidestep the current limitations of the marshaller not marshalling named types.  However, note that the spread operator **_will not solve the case where there are nested types_**.  However, #1533 should be able to solve the nesting issue for named types, if the marshaller allows an option to marshall v2 style.

**Comment by trivikr** [comment]
> note that the spread operator will not solve the case where there are nested types.

Verified that the spread operator will not work on nested interfaces.

<details>
<summary>Example code</summary>

```ts
import { DynamoDB, waitForTableExists } from "@aws-sdk/client-dynamodb";
import { marshall } from "@aws-sdk/util-dynamodb";

interface NameInterface {
  firstname: string;
  lastname: string;
}

interface TableItemInterface {
  id: string;
  name: NameInterface;
}

(async () => {
  const region = "us-west-2";
  const client = new DynamoDB({ region });

  const TableName = `test-${Math.ceil(Math.random() * 10 ** 10)}`;

  const params = {
    TableName,
    AttributeDefinitions: [{ AttributeName: "id", AttributeType: "S" }],
    KeySchema: [{ AttributeName: "id", KeyType: "HASH" }],
    BillingMode: "PAY_PER_REQUEST",
  };

  await client.createTable(params);
  await waitForTableExists({ client, maxWaitTime: 60 }, { TableName });

  const itemContents: TableItemInterface = {
    id: "id",
    name: { firstname: "John", lastname: "Doe" },
  };
  await client.putItem({
    // TS error: Type 'NameInterface' is not assignable to type 'NativeAttributeValue'.
    Item: marshall({ ...itemContents }),
    TableName,
  });

  await client.getItem({
    Key: marshall({ id: itemContents.id }),
    TableName,
  });

  await client.deleteTable({ TableName });
})();
```

</details>

**Comment by trivikr** [comment]
The solution given in https://github.com/microsoft/TypeScript/issues/15300#issuecomment-530667695 works with simple interfaces, but not nested ones.
[TypeScript Playground](https://www.typescriptlang.org/play?target=7#code/KYDwDg9gTgLgBDAnmYcAqwDOMBiBXAOwGMBJAsPeAXgCg44AfObKASwIHM7G4BvOANoBrYIgBczGG04BdCS3Yc4AXwDcNGgBNgRADYBDKKiIQC2BFlyEiEgDxo4oGMAKbMfAQGk47OCMQQAGbocuiW+MRkFDDKAHwAFBAARgBWOjASaACUcFSxcABuEKya6jTszlCB+kSoJADKUnzcgRAQ8lKK6soa2nqGxqbmrI1QEg1S6gD0U3AAchBwAKJQUNA0ztgRRPEjUlll5QSV1bVwJAAKwFCYps30JR3SHOr0BPoAtsASvNz0gawbjB3l8nl0-nADNgQd9JM9XioaD0tDoDEY4CYzPBWFcbqZxrjbgR1JsrMRdoTTAcNBVrqc6qMAIKuOZ4D73OCtdpw8H0JKGCQENlJa7dXqogYYobYpkstnjWWaVkfaazNDILBwIJwMBrFCwRBwADk-KgRrgkvYJg+YH0MFYSV0wAAdOgNcahR8RWafO4CBB4PpMJhWBx3o7UDBFkgUMaFJwjc6NuFrLtFcrqTTjnSanVKQQ5p9UL9-oDoUWwZwEVDgRWeVWkUcTrnzvm5pZgJoOY96y9uDCCdciYWvmKUf10Zjhm2O5pB3iC7PVW6UO5tbqIPqkB6i+bLcQIDa7Q6nUmZvR1bGjZch6YR8Bzaw-QGLcHQ+GnQho+6jRgttYokoRMaHPegV1Qa8213X04H9QM3zDfQIy-BAf34YRRErDhQnjF4VGA0ltgpW9F2wTsDiAA)

We can still implement it to allow passing non-nested interfaces as an input to marshall function.

**Comment by trivikr** [comment]
An option to support nested interfaces would be to advise users to extend their interfaces from a generic interface like this:

```ts
import { NativeAttributeValue, INativeAttributeValue } from "@aws-sdk/util-dynamodb";

interface IPersonName extends INativeAttributeValue<NativeAttributeValue> {
  firstname: string;
  lastname: string;
}
interface IPersonNested extends INativeAttributeValue<NativeAttributeValue> {
  id: string;
  name: IPersonName;
}
```

where:
```ts
interface INativeAttributeValue<T> {
  [key: string]: T
}
```

[TypeScript playground](https://www.typescriptlang.org/play?target=7#code/C4TwDgpgBAKhDOwBiBXAdgYxuaBeAUFFAD5SIBOAlmgOaElQDeUA2gNYQgBcZwVtAXR4VqNKAF8A3PnwATCBgA2AQ3LQMAezSIowBMnQYeAHhhQIADz1pZ8JiwDSUalA4gNAM1hDY+1JmxIcQA+AAoNACMAKwVgHhgASihcYKgANw1KWWl8aj1yD2UMaDhEfwwASTR8wuLTVMZ6dk5hPlEfGHxxXOqIAqLoCoAFPvgtADllAFtoS2tbXzLDKpqB4xFaBvoPSnJENGmIVv4aaSIVfcPj0WluvL7awZG9if0IWSZ6LOvaM6gDmY8YajCaHW4yeRKVTqLQ6SjPMZocZvWRAhGvRDvaQAemxUHGGnM5HIGnI+D0S0woXhIKRKISkiAA)

This way, the customer code would be guaranteed to work with marshall function in `@aws-sdk/util-dynamodb`
What do you think @stang-tgs @vivmaha ?

**Comment by trivikr** [comment]
A draft PR to accept interface as an input to marshall is posted at https://github.com/aws/aws-sdk-js-v3/pull/1952

**Comment by stang-tgs** [comment]
Hi @trivikr,

I don't think this would be a practical solution and may be in conflict with DynamoDb's NoSQL nature.  That is, strongly typed Typescript vs NoSQL.  By requiring all "marshallable types" to inherit from `INativeAttributeValue` would be an inconvenience to the user, not practical, and code structure will be compromised for developers that decoupled their application data models from database models.  That is, by requiring the inheritance/implementation of the `INativeAttributeValue`, both the database layer as well as the higher application layer MUST know the "shape" of all objects, including the nested types.

Using your example, what if I wanted I wanted to add a `pets` property to an `IPerson` such that:

```ts
interface IPersonNested extends INativeAttributeValue<NativeAttributeValue> {
  id: string;
  name: IPersonName;
  pets: IPet[];
}
```
Perhaps the database layer doesn't care what the "shape" of `IPet` is, just that it's a list/array of "something", allowing the application layer to have free reign to define what `IPet` is without the database layer knowing what the "shape" of `IPet` is.

This brings me to rewind the conversation a bit and ask, what is really the purpose of, and is it really worth it:
![image](https://user-images.githubusercontent.com/32852813/105610779-41633480-5d66-11eb-9f3a-221b0fc8a7a5.png)
in [convertToAttr.ts](https://github.com/aws/aws-sdk-js-v3/blob/master/packages/util-dynamodb/src/convertToAttr.ts#L17)?

@trivikr mentioned:
>@vivmaha The DocumentClient in v2 had several bugs. For example, it stored a Date object as a map.
The types in marshall are narrower to avoid bugs like these.

Requiring the user to define the shapes of all data that is to be stored to the database by inheriting from this interface is, in my humble opinion, probably worse than telling the user that the marshaller will marshall all `object` types as Dynamo `Map`.  In other words, I believe that https://github.com/aws/aws-sdk-js-v3/blob/master/packages/util-dynamodb/src/convertToAttr.ts#L12-L22 should really look like:

```ts
export const convertToAttr = (data: NativeAttributeValue, options?: marshallOptions): AttributeValue => {
  if (Array.isArray(data)) {
    return convertToListAttr(data, options);
  } else if (data instanceof Set) {  //CHANGED TO INSTANCEOF (debatable whether this is correct or not)
    return convertToSetAttr(data as Set<any>, options);
  } else if (typeof data === "object") { //CHANGED TO TYPEOF (yes, this will break binary marhsalling, but can be fixed with premptive checks)
    return convertToMapAttr(data as { [key: string]: NativeAttributeValue }, options);
  } else {
    return convertToScalarAttr(data as NativeScalarAttributeValue, options);
  }
};
```

Notice that with this implementation, the marshaller will work if I tried to marshall:
```ts
class MySet extends Set {}
```
whereas the current implementation won't.

I'm aware that the above implementation would break marshalling of binary types, since they would just look like `object`s, but if you want the marshaller to help marshall binary, you can put in binary type checks before the `object` type check.  I think all other Scalars should be covered?

For the generic `object`, I don't believe the narrow protections actually benefit the user.  Using serializing `Date` as an example, in  my personal use case, for the v2 DocumentClient, we know all objects are serialized to maps, so we are cognize that by the time our models are constructed and passed to the database layer, our Dates must be transformed into a number before constructing our model.  Whether the developer chooses to use `Date` above the database layer is really up to developer.  In this case, I do not think it is the Dynamo library's responsibility to "strictly protect the user".  In fact, while writing this and reviewing the marshalling code, I see the marshaller is attempting to protect the user from storing numbers that are in range of `Number.MIN_SAFE_INTEGER` and `Number.MAX_SAFE_INTEGER`.  Again, I believe the marshaller is being overprotective and presumptuous, in that, how does the marshaller know I didn't want a double precision number?  Perhaps I don't want to use a bigint because this field might sometimes contain small fraction numbers.

In any case, getting the marshaller "exactly correct" will be impossible, because JavaScript/TypeScript types do not map one to one.  Because of the asymmetry, there will be some "rules" of how the marshaller works;  but I believe the marshaller should have a minimal set of rules and protections, and leave the stricter rules for the developer to define, because the SDK can't assume/know how the developer wants to store the data.

Sorry for the long winded post, but I hope this explains my thoughts.

**Comment by trivikr** [comment]
The support to convert class instances to map was added in https://github.com/aws/aws-sdk-js-v3/pull/1969, and released in `v3.4.0`

The marshall function now accepts TypeScript interfaces.
Closing this issue as fixed, please reopen or create new issue if it persists.
